### ISSUE TITLE
yum module doesn't call rhnplugin, because module-file is named yum

### ISSUE THREAD

**Issue by luto** [ansible/ansible#13527]
When a RHEL/CentOS server is managed via spacewalk/rhn/satelite, yum uses `yum-rhn-plugin` to get meta-data from the spacewalk server, instead of querying the repos directly.

Example task:

``` yaml
- yum: name=curl disablerepo=* enablerepo=centos-7-x86_64-cityfan
```

(`cityfan` => http://www.city-fan.org/ftp/contrib/yum-repo/rhel7/x86_64/)

Generated output:

``` json
{
  "msg": "Error setting/accessing repos: Error getting repository data for centos-7-x86_64-cityfan-curl-only, repository not found",
  "failed": true
}
```

[`rhnplugin.py#L94-L104`](https://github.com/spacewalkproject/spacewalk/blob/4d2d833a1d5baf92a4c01883e04ac2aa9f57aa0e/client/rhel/yum-rhn-plugin/rhnplugin.py#L94-L104) checks, if the basename of the invoked command is `yum` and if there are no parameters (`cmd_args == []`).

The problem is that ansible executes the yum-module under a path similar to `/home/ansible/.ansible/tmp/ansible-tmp-144929.24-4078098/yum`. So the basename is `yum` and there are no parameters, which causes `yum-rhn-plugin` to never actually query the server. The repolist stays empty, `centos-7-x86_64-cityfan` is never found and the task fails here [yum.py#L926](https://github.com/ansible/ansible-modules-core/blob/74609ce42cb140dba528fee18422467a32acd995/packaging/os/yum.py#L926), with `enablerepo` being the repo name(s).

Spacewalk/yum isn't the only project to check, if it's executed in a certain way by checking its `argv[0]`. So, in order to fix this, I'd suggest to randomize the executable name, just like the path to the dropped ansible-module python-file: `yum-6812`.

What are your thoughts on this?

**Comment by abadger** [comment]
I was pointed at this today and took a look.  This should be fixed since 2.1.

2.1.0 introduced AnsiBallZ as the method of constructing the new-style ansible modules which ship with ansible.  This method of constructing the ansible modules places the module into a file named ansible_module_{module_name}.py and then, on the remote machine, executes that file.  So from 2.1 onwards, the module is not executed as "yum"; instead it is executed as "ansible_module_yum.py".  This should fix the problem raised here.  If you want to see how this works, you can read an introduction here: http://docs.ansible.com/ansible/dev_guide/developing_program_flow_modules.html#ansiballz

and follow the instructions on debugging modules here to see the actual files in play: http://docs.ansible.com/ansible/dev_guide/developing_modules.html#debugging-ansiblemodule-based-modules


Non new-style modules do not have this rewritten filename so in 2.1 they still have this issue.  However, in 2.2 we added support for binary modules.  On Windows, binary modules must have a .exe file extension.  So we made all modules shipped over the wire use the filenames that they have on disk.  So if a module were to be named yum.py on disk,  It would be executed on the remote machine as "yum.py", not as "yum".  So this should solve the problem for non-newstyle modules as well.  The only difficulty is if someone creates a site-specific custom module without a file extension.  But in that case, they can just rename the file to have a file extension and it should all work.

Sorry it took so long to reply to this... with over 2000 open tickets, sometimes the right people don't get to see the tickets that they have the answers for right away.

I'll leave this open for a few days in case you're still having an issue with this in current ansible and then I'll close it since it should be fixed.

**Comment by luto** [comment]
Looks good now. I'm no longer working on the code base which needed this, so I sadly can't test my original issue. Thank you for the update!
